### DISCUSSION TITLE
Action for Paste Selection

### DISCUSSION THREAD

**Discussion by lilyball** [ghostty-org/ghostty#2670] [Feature Requests, Ideas] [answered]
One of my favorite Terminal.app features is it has a menuitem ⇧⌘V for Paste Selection. What this does is whatever the last text I selected in Terminal.app was, it pastes that. So this is kind of like `copy-on-select` behavior except it stores it in a separate pasteboard and only pasted when I use the Paste Selection action. This can be used across windows too, and it's also persisted across relaunch. I'd really love to see Ghostty pick up the same feature.

Implementation note: my initial guess here is Terminal.app uses `+[NSPasteboard pasteboardWithName:]` to create a private pasteboard for this purpose, which would also explain how it persists it (since I don't see it writing the string to preferences). I don't know what name it's using for this. If Ghostty does this as a macOS-only feature then it could perhaps do the same thing.

**Comment by mitchellh** [comment]
This is actually already supported in Ghostty, just not on macOS because I didn't realize you could make secondary named pasteboards. This is implemented because Linux has this natively and Linux users utilize this heavily (middle click always pastes from the selection clipboard on Linux). I have to do some more research but I think we could support this with minimal effort (mostly Swift work). 

Note on Linux, the selection clipboard also persists across restarts because it is connected to the desktop. Selecting text in another application overwrites it (it's global). So this is slightly different from how macOS behave. But from the way you describe it, Terminal.app seems to be mostly emulating this Linux feature within just a single application.

**Reply by lilyball** [reply]
Yeah, I mean I haven't tried to prove that this is what Terminal.app does beyond the fact that I ran `strings` on the binary and found `pasteboardWithName:` and objdump finds a method `-[TTView selectionPasteboard]`, but it makes a lot of sense. I don't know what name Terminal gives to the pasteboard but I imagine you could just come up with something sufficiently unique (e.g. perhaps including bundle ID). I've also never tried making custom pasteboards this way myself, I just know the documentation says it's supported.

**Reply by lilyball** [reply]
Hah, I found it! Terminal does exactly what I just suggested you do, it names the pasteboard using its bundle ID (`com.apple.Terminal.selection`), and other apps can read it too by using the same name. Not that I think Ghostty should share the Terminal selection pasteboard, that would be a little surprising, but `com.mitchellh.ghostty.selection` would be a good name to use.

**Reply by mitchellh** [reply, marked as answer]
#2722

**Reply by poige** [reply]
So basically just because of some-one uses Terminal.app's not that widely known feature now all macos users of Ghostty would be struggling to realise what the heck has happened to copy-on-select which seemingly ain't working anymore — right? )

**Reply by albertinix** [reply]
Agreed, such a weird decision. `copy-on-select` should automatically copy to the general clipboard. That should be default. **Other** things are nice to have, sure, but shouldn't override the basic behaviour that most people expect.
